File: Assets/Scripts/FootSteps.js
// Footsteps.js
// Handles all of the footstep sounds for the player.

var audioStepLength = 0.3;
var footAudio : AudioClip[];
var controller : CharacterController;
private var lastPos : Vector3 = Vector3(0,0,0);
function Start() {
	PlayStepSounds();
}

function PlayStepSounds() {
    while (this.enabled == true) {
		if (transform.position == lastPos) {
			audio.Stop();
			yield;
		} else if (controller.isGrounded && controller.velocity.magnitude > 0.3) {
			//var volume = Mathf.Clamp01(Time.time);
			audio.clip = footAudio[Random.Range(0, footAudio.Length)];
			audio.Play();
			lastPos = transform.position;
			//audio.PlayOneShot(footAudio[Random.Range(0, footAudio.Length)], volume);
			yield WaitForSeconds(audioStepLength);
		} else {
			audio.Stop();
			yield;
		}
    }
	audio.Stop();
}

File: Assets/Scripts/Footsteps.cs
using UnityEngine;
using System.Collections;

public class Footsteps : MonoBehaviour {
	
	public AudioClip[] footAudio;
	public AudioClip[] landAudio;
	
	public GameObject soundOrb;
	
	public float intensityDecrease = .4f;
	public float minStepIntensity = .15f;
	public float maxStepIntensity = 1;
	
	public float minLandVal = 4;
	public float maxLandVal = 6.25f;
	public float minLandIntensity = .8f;
	public float maxLandIntensity = .95f;
	
	public float stepSoundOrbMaxPower = 3f;
	public float landSoundOrbMaxPower = 6f;
	
	public void Footstep(float intensity) {
		intensity -= intensityDecrease;
		if (intensity > maxStepIntensity){
			intensity = maxStepIntensity;
		}
		if (intensity < minStepIntensity){
			intensity = minStepIntensity;	
		}
		audio.volume = intensity;
		audio.clip = footAudio[Random.Range(0, footAudio.Length)];
		audio.Play();
		
		makeSound (intensity*stepSoundOrbMaxPower);
	}
	
	public void Land(float downVelocity) {		
		downVelocity = Mathf.Abs(downVelocity);
		
		float intensity = (downVelocity-minLandVal)/(maxLandVal-minLandVal);
		if (intensity < 0){
			intensity = 0;
		}
		if (intensity > 1){
			intensity = 1;	
		}
		
		audio.volume = minLandVal+(intensity*(maxLandVal-minLandVal));
		audio.clip = landAudio[Random.Range(0, landAudio.Length)];
		audio.Play();

		makeSound (intensity*landSoundOrbMaxPower);
	}
	
	void makeSound (float intensity){
		if (intensity > 0){
			soundOrb.GetComponent<SoundOrb>().intensity = intensity;		
			GameObject sound = (GameObject)Instantiate(soundOrb, transform.position, transform.rotation);		
		}
	}
	
}

